json_rpc/client.nim

```diff
@@ -26,13 +26,16 @@ proc getNextId*(client: RpcClient): ClientId =
 proc rpcCallNode*(path: string, params: JsonNode, id: ClientId): JsonNode =
   %{"jsonrpc": %"2.0", "method": %path, "params": params, "id": %id}
 
+proc rpcNotificationNode*(path: string, params: JsonNode): JsonNode =
+  %{"jsonrpc": %"2.0", "method": %path, "params": params}
+
 method call*(client: RpcClient, name: string,
              params: JsonNode): Future[Response] {.
-    base, async, gcsafe, raises: [Defect, CatchableError].} =
+    base, async, gcsafe, raises: [Defect, CatchableError, Exception].} =
```

Comment:
throughout, `Exception` should never appear as part of the raises list - if it "leaks" from somewhere, it should be contained in that place by correctly annotating that source, or worst case, surrounding the code with `try: ... except CatchableError as exc: raise exc except Exception ass exc: raiseAssert exc.msg` - usually Execption infections come from incorrect forward and/or `proc` declarations.

Reply:
I will track down that. I believe it was caused by async when used in combination with `asyncdispatch` but I have to verify that.

Reply:
we generally don't support / work with `asyncdispatch` (it doesn't work in faststreams for example), so that would be surprising - importantly though, raises lists are infectious.

ah, I also see that this function in particular is `async` - those should not have `raises` lists at all (chronos does not yet fully support raises lists for async functions and adding them causes trouble - the correct annotation for an async function is `raises: [Defect]` which should be done with a `push raises: [Defect]` on top of every file

Reply:
FYI, here is the root cause:

``` nim
import std/asyncdispatch

proc echoString(params: string): Future[string]
    {.async, raises: [Defect, CatchableError].} =
  return params
```

Fails with

> /home/yyoncho/Sources/nim/Nim/lib/pure/asyncmacro.nim(232, 33) Error: echoStringNimAsyncContinue_436207620() can raise an unlisted exception: Exception

Reply:
yes, so what `async` does is turn the function into two functions - one "wrapper" that creates a future and one closure that contains the body that you wrote - the `raises` annotation is applied to the "wrapper", not the "user code" (as it perhaps should be, but this has other problems), hence the correct `raises` annotation for `async` is `raises: [Defect]`, with async functions - if asyncdispatch fails here, it's a deeper issue with AD that needs to be resolved in the std lib most likely - we require that no exceptions are raised from the "wrapper", and enforce that in most projects (nim-json-rpc is just a little bit behind in maintenance - more info:

* https://status-im.github.io/nim-style-guide/errors.html
* https://github.com/status-im/nim-chronos#exception-effects